# <p align="center">Siemens Sales Forecast</p>

---

## <p align="center">*1 - Exploratory Data Analysis & Pre-processing*</p>

---

### 👥 **Team Members**
- **Ana Farinha** *(Student Number: 20211514)*  
- **António Oliveira** *(Student Number: 20211595)*  
- **Mariana Neto** *(Student Number: 20211527)*  
- **Salvador Domingues** *(Student Number: 20240597)*  

📅 **Date:** *April 1, 2025*  
📍 **Prepared for:** *Siemens*  

**GitHub Repo:** https://github.com/MGN19/Siemens-forecast

---

# ToC

<a class="anchor" id="top"></a>


1. [Import Libraries & Data](#1.-Import-Libraries-&-Data) <br><br>

2. [Data Wrangling](#2.-Data-Wrangling) <br><br>

3. [Data Exploration](#3.-Data-Exploration) <br><br>

# 1. Import Libraries & Data

In [8]:
import pandas as pd

import utils as u

pd.set_option('display.max_columns', None)

**Data**

In [9]:
test = pd.read_csv('data/Case2_Test Set Template.csv', 
                 sep = ';')

sales_data = pd.read_csv('./data/Case2_Sales data.csv', 
                 sep = ';')

market_data = pd.read_excel('./data/Case2_Market data.xlsx', 
                            index_col = 'Unnamed: 0')


# 2. Data Wrangling

<a href="#top">Top &#129033;</a>

- make necessarry changes to datasets to make them usable

## 2.1 Sales Data


In [10]:
sales_data.head(2)

DATE Mapped_GCK Sales_EUR
0  01.10.2018         #1         0
1  02.10.2018         #1         0

**Exploring the dataset structure with .info() method**

The `.info()` method provides more information about the dataset, including the existence of missing values and the data type of the elements of each row.

In [11]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9802 entries, 0 to 9801
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   DATE        9802 non-null   object
 1   Mapped_GCK  9802 non-null   object
 2   Sales_EUR   9802 non-null   object
dtypes: object(3)
memory usage: 229.9+ KB


With the information above, we can see the following:
- all of the columns in this dataset are categorical variables. For analysis purposes, the `DATE` column will be transformed into a date format, and the `Sales_EUR` column into a float. 
- we can also see that there might not be any missing values in any of the columns of the dataset.

Based on this, we will first change the data types of these 2 columns identified so that analysis can be proprerly made.

In [12]:
# convert to datetime
sales_data['DATE'] = pd.to_datetime(sales_data['DATE'], 
                                    format='%d.%m.%Y')

# convvert to numbers
sales_data['Sales_EUR'] = sales_data['Sales_EUR'].apply(lambda x: x.replace(',', '.')).astype(float)

sales_data.head(1)

DATE Mapped_GCK  Sales_EUR
0 2018-10-01         #1        0.0

**Convert to Monthly Data**

In [13]:
# convert to montlhy sales
sales_data["Date"] = sales_data["DATE"].dt.to_period("M")
monthly_sales = sales_data.groupby(["Date", 
                                     "Mapped_GCK"]).agg({"Sales_EUR": "sum"}).reset_index()

monthly_sales.head(7)

Date Mapped_GCK    Sales_EUR
0  2018-10         #1  36098918.79
1  2018-10        #11   1021303.50
2  2018-10        #12     28686.33
3  2018-10        #13     27666.10
4  2018-10        #14      5770.00
5  2018-10        #16    333196.87
6  2018-10        #20      4563.14

Make each Product categ a column

In [14]:
pivoted_data = monthly_sales.pivot(index='Date', 
                                 columns='Mapped_GCK', 
                                 values='Sales_EUR')

pivoted_data.head(3)

Mapped_GCK           #1        #11       #12       #13      #14         #16  \
Date                                                                          
2018-10     36098918.79  1021303.5  28686.33  27666.10   5770.0   333196.87   
2018-11      5140760.00  1898844.8   1070.00  68180.00  17130.0  1377694.32   
2018-12     37889612.12  1226122.0  17880.60  15655.18      0.0  4762524.66   

Mapped_GCK      #20           #3       #36         #4          #5         #6  \
Date                                                                           
2018-10     4563.14   8089465.96   6474.60  397760.69  2499061.19  369231.60   
2018-11     5798.14  11863001.51  21617.61  371322.42  8993944.04  473046.96   
2018-12      918.65   8736859.39  13924.52  430100.96  6947507.31  999472.69   

Mapped_GCK         #8       #9  
Date                            
2018-10     586052.74  3219.32  
2018-11     526292.77  1875.90  
2018-12     271490.71     0.00

## 2.2 Market Data

<a href="#top">Top &#129033;</a>

In [15]:
market_data.head(3)

China  \
Index 2010=100 (if not otherwise noted)  Production Index Machinery & Electricals   
date                                                               MAB_ELE_PRO156   
 2004m2                                                                 16.940704   

                                                                         China.1  \
Index 2010=100 (if not otherwise noted)  Shipments Index Machinery & Electricals   
date                                                              MAB_ELE_SHP156   
 2004m2                                                                16.940704   

                                                                           France  \
Index 2010=100 (if not otherwise noted)  Production Index Machinery & Electricals   
date                                                               MAB_ELE_PRO250   
 2004m2                                                                112.091273   

                                                                        France.1  \
Index 2010=100 (if not otherwise noted)  Shipments Index Machinery & Electricals   
date                                                              MAB_ELE_SHP250   
 2004m2                                                                83.458866   

                                                                          Germany  \
Index 2010=100 (if not otherwise noted)  Production Index Machinery & Electricals   
date                                                               MAB_ELE_PRO276   
 2004m2                                                                 82.623037   

                                                                       Germany.1  \
Index 2010=100 (if not otherwise noted)  Shipments Index Machinery & Electricals   
date                                                              MAB_ELE_SHP276   
 2004m2                                                                79.452532   

                                                                            Italy  \
Index 2010=100 (if not otherwise noted)  Production Index Machinery & Electricals   
date                                                               MAB_ELE_PRO380   
 2004m2                                                                124.289603   

                                                                         Italy.1  \
Index 2010=100 (if not otherwise noted)  Shipments Index Machinery & Electricals   
date                                                              MAB_ELE_SHP380   
 2004m2                                                                86.560493   

                                                                            Japan  \
Index 2010=100 (if not otherwise noted)  Production Index Machinery & Electricals   
date                                                               MAB_ELE_PRO392   
 2004m2                                                                 109.33401   

                                                                         Japan.1  \
Index 2010=100 (if not otherwise noted)  Shipments Index Machinery & Electricals   
date                                                              MAB_ELE_SHP392   
 2004m2                                                               110.495272   

                                                                      Switzerland  \
Index 2010=100 (if not otherwise noted)  Production Index Machinery & Electricals   
date                                                               MAB_ELE_PRO756   
 2004m2                                                                 91.221862   

                                                                   Switzerland.1  \
Index 2010=100 (if not otherwise noted)  Shipments Index Machinery & Electricals   
date                                                              MAB_ELE_SHP756   
 2004m2                                                                89.987275   

        

In [16]:
info = market_data.T.iloc[:, :2]
info

Index 2010=100 (if not otherwise noted)  \
China                      Production Index Machinery & Electricals   
China.1                     Shipments Index Machinery & Electricals   
France                     Production Index Machinery & Electricals   
France.1                    Shipments Index Machinery & Electricals   
Germany                    Production Index Machinery & Electricals   
Germany.1                   Shipments Index Machinery & Electricals   
Italy                      Production Index Machinery & Electricals   
Italy.1                     Shipments Index Machinery & Electricals   
Japan                      Production Index Machinery & Electricals   
Japan.1                     Shipments Index Machinery & Electricals   
Switzerland                Production Index Machinery & Electricals   
Switzerland.1               Shipments Index Machinery & Electricals   
United Kingdom             Production Index Machinery & Electricals   
United Kingdom.1            Shipments Index Machinery & Electricals   
United States              Production Index Machinery & Electricals   
United States.1             Shipments Index Machinery & Electricals   
Europe                     Production Index Machinery & Electricals   
Europe.1                    Shipments Index Machinery & Electricals   
Unnamed: 19                             World: Price of Base Metals   
Unnamed: 20                                  World: Price of Energy   
Unnamed: 21                      World: Price of Metals  & Minerals   
Unnamed: 22                       World: Price of Natural gas index   
Unnamed: 23                      World: Price of Crude oil, average   
Unnamed: 24                                  World: Price of Copper   
Unnamed: 25                               United States: EUR in LCU   
Producer Prices                 United States: Electrical equipment   
Producer Prices.1              United Kingdom: Electrical equipment   
Producer Prices.2                       Italy: Electrical equipment   
Producer Prices.3                      France: Electrical equipment   
Producer Prices.4                     Germany: Electrical equipment   
Producer Prices.5                       China: Electrical equipment   
production index      United States: Machinery and equipment n.e.c.   
production index.1            World: Machinery and equipment n.e.c.   
production index.2      Switzerland: Machinery and equipment n.e.c.   
production index.3   United Kingdom: Machinery and equipment n.e.c.   
production index.4            Italy: Machinery and equipment n.e.c.   
production index.5            Japan: Machinery and equipment n.e.c.   
production index.6           France: Machinery and equipment n.e.c.   
production index.7          Germany: Machinery and equipment n.e.c.   
production index.8              United States: Electrical equipment   
production index.9                      World: Electrical equipment   
production index.10               Switzerland: Electrical equipment   
production index.11            United Kingdom: Electrical equipment   
production index.12                     Italy: Electrical equipment   
production index.13                     Japan: Electrical equipment   
production index.14                    France: Electrical equipment   
production index.15                   Germany: Electrical equipment   

                                       date  
China                        MAB_ELE_PRO156  
China.1                      MAB_ELE_SHP156  
France                       MAB_ELE_PRO250  
France.1                     MAB_ELE_SHP250  
Germany                      MAB_ELE_PRO276  
Germany.1                    MAB_ELE_SHP276  
Italy                        MAB_ELE_PRO380  
Italy.1                      MAB_ELE_SHP380  
Japan                        MAB_ELE_PRO392  
Japan.1                      MAB_ELE_SHP392  
Switzerland                  MAB_ELE_PRO756  
Switzerland.1                MAB_ELE_SHP756  
United Kingdom       

Rename Cols

In [17]:
market_data = market_data.rename(columns = u.rename_dict)

Remove unnecessary rows

In [18]:
market_data = market_data.reset_index()
market_data = market_data[2:]
market_data.rename(columns={'index': 'Date'}, inplace=True)
market_data

Date CHI Production Index CHI Shipments Index FRA Production Index  \
2     2004m2            16.940704           16.940704           112.091273   
3     2004m3            23.711852           23.711852           136.327976   
4     2004m4            24.435235           24.435235           117.791806   
5     2004m5            23.708115           23.708115           109.002541   
6     2004m6            27.009138           27.009138           133.785737   
..       ...                  ...                 ...                  ...   
216  2021m12           310.763183          310.763183           100.565744   
217   2022m1           235.956129          235.956129            85.743503   
218   2022m2           235.956129          235.956129             90.60354   
219   2022m3           329.413367          329.413367           107.843548   
220   2022m4           267.373145          267.373145             87.69811   

    FRA Shipments Index GER Production Index GER Shipments Index  \
2             83.458866            82.623037           79.452532   
3            106.168192           100.556582           97.012918   
4             92.007646            89.653203           84.932358   
5             85.696486            86.880571           82.372794   
6            106.641482            99.010814            95.10874   
..                  ...                  ...                 ...   
216          134.589504           118.103281          149.364286   
217           108.15632             94.55061          120.353403   
218           117.71577           103.987916          129.383676   
219           136.85872           121.308119          151.201314   
220          116.528738            99.522205          127.022869   

    ITA Production Index ITA Shipments Index JAP Production Index  \
2             124.289603           86.560493            109.33401   
3             143.411662          106.344544           140.884616   
4             129.083828           95.579673           105.853579   
5             135.590391          100.087039           101.864777   
6             136.424935          110.889719            120.33292   
..                   ...                 ...                  ...   
216            94.006826          150.482735           127.771735   
217            86.851008          101.258277           110.460181   
218           106.583758          120.956538           117.879631   
219           124.637966          153.645142           152.000561   
220            103.55669          128.733305           114.262328   

    JAP Shipments Index SWI Production Index SWI Shipments Index  \
2            110.495272            91.221862           89.987275   
3            144.686166            85.866287           79.883583   
4            102.655769            85.622508           79.740802   
5            100.305285            85.378729           79.598021   
6             119.61638             85.13495           79.455239   
..                  ...                  ...                 ...   
216          131.029703           106.704029          104.819189   
217          110.823532            103.49926           101.70157   
218          118.300232           100.294492           98.583952   
219          156.400634            97.089723           95.466333   
220          115.012049                  NaN                 NaN   

    UK Production Index UK Shipments Index USA Production Index  \
2            111.353812          73.601265             107.6014   
3            127.558608          84.047595           110.187364   
4            108.732297          73.026027           108.166564   
5              110.6452          74.591883           108.425887   
6             122.02096          82.343346           110.569933   
..                  ...                ...                  ...   
216          101.273544                NaN           107.040766   
217           95.003541                NaN           111.052133   
218  

Convert dates to datetime

In [19]:
# convert to datetime
market_data["Year"] = market_data["Date"].str.extract(r"(\d{4})").astype(int)
market_data["Month"] = market_data["Date"].str.extract(r"m(\d{1,2})").astype(int)

market_data["Date"] = pd.to_datetime(market_data["Year"].astype(str) + "-" + market_data["Month"].astype(str), format="%Y-%m").dt.to_period("M")

market_data.drop(['Year', 'Month'], axis = 1, inplace = True)

Convert numbers to float

In [20]:
for column in market_data.columns:
    if column in ['Date', 'Month', 'Year']:
        continue
    else:
        market_data[column] = market_data[column].astype(str).str.replace(',', '.', regex=False)
        market_data[column] = pd.to_numeric(market_data[column], errors='coerce')

In [21]:
market_data.set_index('Date', inplace=True)

In [22]:
market_data.head(4)

CHI Production Index  CHI Shipments Index  FRA Production Index  \
Date                                                                       
2004-02             16.940704            16.940704            112.091273   
2004-03             23.711852            23.711852            136.327976   
2004-04             24.435235            24.435235            117.791806   
2004-05             23.708115            23.708115            109.002541   

         FRA Shipments Index  GER Production Index  GER Shipments Index  \
Date                                                                      
2004-02            83.458866             82.623037            79.452532   
2004-03           106.168192            100.556582            97.012918   
2004-04            92.007646             89.653203            84.932358   
2004-05            85.696486             86.880571            82.372794   

         ITA Production Index  ITA Shipments Index  JAP Production Index  \
Date                                                                       
2004-02            124.289603            86.560493            109.334010   
2004-03            143.411662           106.344544            140.884616   
2004-04            129.083828            95.579673            105.853579   
2004-05            135.590391           100.087039            101.864777   

         JAP Shipments Index  SWI Production Index  SWI Shipments Index  \
Date                                                                      
2004-02           110.495272             91.221862            89.987275   
2004-03           144.686166             85.866287            79.883583   
2004-04           102.655769             85.622508            79.740802   
2004-05           100.305285             85.378729            79.598021   

         UK Production Index  UK Shipments Index  USA Production Index  \
Date                                                                     
2004-02           111.353812           73.601265            107.601400   
2004-03           127.558608           84.047595            110.187364   
2004-04           108.732297           73.026027            108.166564   
2004-05           110.645200           74.591883            108.425887   

         USA Shipments Index  Europe Production Index  Europe Shipments Index  \
Date                                                                            
2004-02            79.240230                97.122911               80.098530   
2004-03            98.619024               113.783904               96.015929   
2004-04            89.774031               101.715199               85.167236   
2004-05            87.463813               101.275727               84.485767   

         (W) Price of Base Metals  (W) Price of Energy  \
Date                                                     
2004-02                 54.039811            44.123338   
2004-03                 54.666162            47.588957   
2004-04                 54.872715            47.779013   
2004-05                 51.230356            53.590898   

         (W) Price of Metals & Minerals  (W) Price of Natural gas index  \
Date                                                                      
2004-02                       48.747945                       87.076974   
2004-03                       49.256157                       87.192705   
2004-04                       49.423751                       91.379923   
2004-05                       46.468392                       99.044520   

         (W) Price of Crude oil, average  (W) Price of Copper  \
Date                                                            
2004-02                        39.639458            36.623832   
2004-03                        42.592034            39.931055   
2004-04                        42.650637            39.134854   
2004-05                        47.517121            36.278433   

         USA EUR to LCU Conversion Rate   USA EE Production Prices  \


# 3. Data Exploration

<a href="#top">Top &#129033;</a>

## 3.1 Initial Sales Data Exploration

<br>

**Statistics**

By using the `describe()` method we get different statistical measurements, such as mean and standard deviation, for the various features. We will include the parameter 'all' so both statistics for numerical and categorical data appear. 

In [24]:
pivoted_data.describe().T

count          mean           std         min           25%  \
Mapped_GCK                                                                
#1           43.0  3.591177e+07  5.905117e+06  5140760.00  3.440896e+07   
#11          43.0  1.532589e+06  9.811598e+05    18200.00  9.035992e+05   
#12          43.0  1.925462e+05  1.410168e+05     1070.00  3.923521e+04   
#13          43.0  2.346877e+04  1.614016e+04     2550.31  1.046647e+04   
#14          43.0  1.148423e+04  1.611617e+04    -2851.45  2.015025e+03   
#16          43.0  4.277018e+05  7.824899e+05    40360.08  1.375541e+05   
#20          43.0  1.821555e+03  2.307900e+03        0.00  3.308100e+02   
#3           43.0  1.240566e+07  2.577332e+06  3804319.74  1.105907e+07   
#36          43.0  2.322358e+04  4.265128e+04      674.00  6.244495e+03   
#4           43.0  3.634234e+05  1.557109e+05    93226.32  2.797995e+05   
#5           43.0  9.990585e+06  3.731564e+06  1405295.36  7.703432e+06   
#6           43.0  4.800440e+05  2.422862e+05   -15202.96  3.160555e+05   
#8           43.0  8.109586e+05  5.436111e+05   111791.34  3.749379e+05   
#9           43.0  7.173175e+03  8.231269e+03    -4358.10  1.626700e+03   

                    50%           75%          max  
Mapped_GCK                                          
#1          37323903.07  3.888867e+07  44483013.86  
#11          1226122.00  2.161184e+06   3891447.76  
#12           172712.88  3.217950e+05    445648.06  
#13            20663.64  3.096641e+04     68180.00  
#14             5951.85  1.530589e+04     76161.44  
#16           224501.90  3.283725e+05   4762524.66  
#20              842.93  2.543020e+03      8485.60  
#3          12317479.75  1.377567e+07  18686819.85  
#36            12377.77  2.199505e+04    253519.04  
#4            329430.96  4.281612e+05    829442.33  
#5           9232237.68  1.180077e+07  19524640.52  
#6            499147.73  6.119871e+05    999472.69  
#8            599831.93  1.092310e+06   2655036.78  
#9              4488.09  9.583200e+03     39510.75

## 3.2 Initial Market Data Exporation

In [ ]:
market_data.info()

With the information above, we can see the following:
- `date` could be set as index, since there is no duplicated date
- all of the columns in this dataset are stored as categorical variables, except `date`. Therefore, some modifications regarding data types might happen, because these columns' values represent prices and would be more appropriate to store as floats.
- we can also see that there are missing values on `SWI_MC_EL_PROD`, `SWI_MC_EL_SHIP`, `UK_MC_EL_SHIP`, `USA_MC_EL_SHIP`, `UK_EE_PRODUCER_PRICE`, `FRA_EE_PRODUCER_PRICE`, `CHI_EE_PRODUCER_PRICE`, `CHI_MC_EQ_PROD`, `WRL_EE_PROD` and `SWI_EE_PROD`. However, these missing values might be due to the fact that not all countries have, for instance, producers.

We will proceed to change the datatypes of all columns, except `date`, to float and set `date` as index.

**Statistics**

In [26]:
market_data.describe()

CHI Production Index  CHI Shipments Index  FRA Production Index  \
count            219.000000           219.000000            219.000000   
mean             138.303637           138.303637            104.431918   
std               78.883209            78.883209             18.918529   
min               16.940704            16.940704             50.756680   
25%               68.477740            68.477740             93.613505   
50%              133.507690           133.507690            102.736556   
75%              198.473934           198.473934            114.090851   
max              329.413367           329.413367            152.743402   

       FRA Shipments Index  GER Production Index  GER Shipments Index  \
count           219.000000            219.000000           219.000000   
mean            105.316814            107.499126           114.898377   
std              12.762209             11.861942            17.091571   
min              64.420676             74.332913            71.787161   
25%              97.452819            100.560897           103.149778   
50%             106.012166            108.992290           117.428836   
75%             115.030479            115.735786           127.112220   
max             136.858720            130.869962           151.297092   

       ITA Production Index  ITA Shipments Index  JAP Production Index  \
count            219.000000           219.000000            219.000000   
mean             105.228363           105.735378            111.948146   
std               23.509638            19.948183             15.489336   
min               34.213427            45.191710             67.531940   
25%               94.335162            95.985839            103.740049   
50%              105.088474           107.695805            111.683015   
75%              117.031701           119.836360            121.402653   
max              153.940791           153.645142            153.898678   

       JAP Shipments Index  SWI Production Index  SWI Shipments Index  \
count           219.000000            218.000000           218.000000   
mean            112.670602             97.834543            94.784942   
std              16.891947              8.241523             9.153899   
min              64.372344             77.801503            74.639253   
25%             103.453182             92.410183            88.221030   
50%             112.597293             97.573131            94.376831   
75%             121.498141            103.877063           100.736473   
max             159.495942            116.674608           115.321078   

       UK Production Index  UK Shipments Index  USA Production Index  \
count           219.000000          201.000000            219.000000   
mean            108.752949           95.957072            109.418255   
std              12.096725           12.946170              7.891270   
min              61.048022           57.462935             85.994448   
25%             100.498418           86.199717            104.599964   
50%             108.911029           97.868918            110.153555   
75%             118.001772          105.269576            114.585399   
max             137.682590          126.338522            126.650773   

       USA Shipments Index  Europe Production Index  Europe Shipments Index  \
count           218.000000               219.000000              219.000000   
mean            114.160028               108.779990              110.551132   
std              14.633380                11.839462               15.727859   
min              79.240230                69.786633               71.158884   
25%             104.596526               101.729625               98.588003   
50%             115.713379               110.383833              111.954128   
75%             123.419976               115.832526              121.613878   
max             165.926217               134.216163              158.055622 

Based on this, we get the following:

1. **Production and Shipments in Machinery & Electricals**  
   - China’s production and shipment index show high variability, indicating for instance fluctuations in demand, supply chain issues, or policy impacts.
   - At some point in time, Production and Shipments was over 3 times higher than in 2010 (229% increase)
   China was producing 38% more than in 2010 and in terms of other countries it was always higher.
   - Switzerland has the lowest mean production and shipment index, compared to the other countries.
   - UK tends to produce more than export
   - Germany’s shipment index mean (114.9) is higher than its production mean (107.5), which could indicate a focus on exports or strong international demand.

3. **Material Prices**  
   - Gas prices (mean: 103.37, std: 41.74) show high volatility, likely driven by geopolitical factors and global supply-demand shocks.  
   - Copper and metal price have a lower mean (86.34) than other metals.
   - Oil prices have a high range (26.62–168.05), reflecting the impact of market cycles and demand shifts.  

4. **Producer Prices vs. Production Indices (Electrical Equipment)**  
   - Germany has the highest producer price mean (103.17), closely followed by USA (103.17) and Italy (102.16). This suggests that companies in these countries are paying more to produce goods.  
   - Switzerland’s mean producer price is lower than others (97.68).
   - On average, UK had produce more than other countries (116), followed by the USA (110).

5. **Machinery & Equipment Production Trends**  
   - Japan has the highest production mean (115.08), followed by Germany (110.90) and the U.S. (110.42), indicating a strong machinery sector production.  
   - Switzerland has the lowest production indexes (on average 99.7).

# 4. Feature Exploration

<a href="#top">Top &#129033;</a>

## 4.1 Univariate Analysis

**Sales Data**

**Market Data**

## 4.2 Bivariate Analysis

**Sales Data**

**Market Data**

# 5. Data Cleaning & Preprocessing

## 5.1 Duplicates

**Sales Data**

**Market Data**

## 5.2 Missing Values

**Sales Data**

**Market Data**

## 5.3 Feature Engineering

**Sales Data**

**Market Data**

## 5.4 Outliers

**Sales Data**

**Market Data**